sdk/core/azure-core/azure/core/polling/_async_poller.py

```diff
@@ -71,15 +91,95 @@ async def async_poller(client, initial_response, deserialization_callback, polli
     """
+    poller = AsyncLROPoller(client, initial_response, deserialization_callback, polling_method)
+    return await poller
 
-    # This implicit test avoids bringing in an explicit dependency on Model directly
-    try:
-        deserialization_callback = deserialization_callback.deserialize
-    except AttributeError:
-        pass
 
-    # Might raise a CloudError
-    polling_method.initialize(client, initial_response, deserialization_callback)
+class AsyncLROPoller(Awaitable, Generic[PollingReturnType]):
+    """Async poller for long running operations.
+
+    :param client: A pipeline service client
+    :type client: ~azure.core.PipelineClient
+    :param initial_response: The initial call response
+    :type initial_response:
+     ~azure.core.pipeline.transport.HttpResponse or ~azure.core.pipeline.transport.AsyncHttpResponse
+    :param deserialization_callback: A callback that takes a Response and return a deserialized object.
+                                     If a subclass of Model is given, this passes "deserialize" as callback.
+    :type deserialization_callback: callable or msrest.serialization.Model
+    :param polling_method: The polling strategy to adopt
+    :type polling_method: ~azure.core.polling.AsyncPollingMethod
+    """
+
+    def __init__(
+            self,
+            client: Any,
+            initial_response: Any,
+            deserialization_callback: Callable,
+            polling_method: AsyncPollingMethod[PollingReturnType]
+        ):
+        self._polling_method = polling_method
+        self._done = False
+
+        # This implicit test avoids bringing in an explicit dependency on Model directly
+        try:
+            deserialization_callback = deserialization_callback.deserialize # type: ignore
+        except AttributeError:
+            pass
+
+        self._polling_method.initialize(client, initial_response, deserialization_callback)
+
+    def continuation_token(self) -> str:
+        """Return a continuation token that allows to restart the poller later.
+
+        :returns: An opaque continuation token
+        :rtype: str
+        """
+        return self._polling_method.get_continuation_token()
+
+    @classmethod
+    def from_continuation_token(
+            cls,
+            polling_method: AsyncPollingMethod[PollingReturnType],
+            continuation_token: str,
+            **kwargs
+        ) -> "AsyncLROPoller[PollingReturnType]":
+        client, initial_response, deserialization_callback = polling_method.from_continuation_token(
+            continuation_token, **kwargs
+        )
+        return cls(client, initial_response, deserialization_callback, polling_method)
 
-    await polling_method.run()
-    return polling_method.resource()
+    def status(self) -> str:
+        """Returns the current status string.
+
+        :returns: The current status string
+        :rtype: str
+        """
+        return self._polling_method.status()
```

Comment:
Not necessary for this PR - but it would be good if we could open up the contract regarding the expected return type status in future.

Reply:
I've started creating an `azure-core-typing` package multiple times that would include all the protocols that we have (LRO, item paged, page paged etc.). I will take another stab at it as part of the next round of azure client library python design guideline updates.